### ISSUE TITLE
Continuation of Block Style Variations

### ISSUE THREAD

**Issue by mtias** [WordPress/gutenberg#7551]
Once #7362 lands, we'll have the ability to register new block variations (based on classes) with automated previews. With this foundation there are further ideas of how this functionality could be extended.

### Register New Styles

Themes / Plugins should be able to _easily register_ new variations from the client or server.

### Unregister Existing Styles

Themes / Plugins should be able to _unregister_ existing block style variations for a specific block. (A plugin may also want to disable style variations entirely across all blocks.)

### Change the Default

A Plugin / Theme might want to make a specific variation the default one for that block type.

### User Defaults

A user might want to make a specific variation the default whenever they insert a new block of the type.

### Custom User Styles

A user might be able to write a custom class (in the "Advanced" block tab) and register it as a block variation.

Concept:
https://user-images.githubusercontent.com/548849/41718061-1a63de7e-755c-11e8-9680-c6b180730b2a.png

---

We won't be necessarily doing all of these for 5.0 but it's good to discuss them in concert.

**Comment by SuperGeniusZeb** [comment]
>A user might be able to write a custom class (in the "Advanced" block tab) and register it as a block variation.
>
>Concept:
![image](https://user-images.githubusercontent.com/548849/41718061-1a63de7e-755c-11e8-9680-c6b180730b2a.png)

This reminds me a lot of the classes feature in Oxygen Builder:
https://oxygenbuilder.com/documentation/styling-options/classes/

I think this is a great idea! :+1:

**Comment by felixarntz** [comment]
> Themes / Plugins should be able to easily register new variations from the client or server.
> [...]
> Themes / Plugins should be able to unregister existing block style variations for a specific block. (A plugin may also want to disable style variations entirely across all blocks.)

Those two parts are the steps I consider the most crucial for now out of the list. Many themes provide elements in different variations, there may be different separators, quotes, or especially buttons, where you often have a distinction between a regular (secondary) button and a primary button, just like WordPress itself has in the admin. Having this map to Gutenberg in a simple way by registering a block style is an easy-to-implement solution for the theme developer.

In terms of implementation, I can see it either being handled via theme support (as in `add_theme_support( 'block-styles', ... )` and `add_theme_support( 'disabled-block-styles', ... )`, or via extra functions made available via Gutenberg (as in `register_block_styles()` and `unregister_block_styles()`). I think the first is most in line how we would commonly handle it, and I like to reuse the concept of theme support - especially if we only allow adding new styles, but not disabling existing ones, I think we should go that route. On the other hand, this is (or might evolve into) something more specific than just a simple on/off feature or a list of things, so I also see an advantage for introducing specific functions for that purpose.

This was previously discussed in https://github.com/WordPress/gutenberg/issues/5360#issuecomment-391301305 (theme support) and https://github.com/WordPress/gutenberg/issues/5360#issuecomment-385824484 (extra functions), some input around that to consider as well.

**Comment by boemedia** [comment]
I just installed 3.2.0. Is it correct that the block styles/settings have been taken out of the block itself? What's the main reason for it? From a design/accessibility perspective, this is an odd choice. For instance, for people with tunnel vision, the proximity of related actions is very important. Now when you select a block at the bottom of your screen, you won't see what happens at the top left.

**Comment by mtias** [comment]
What do you mean out of the block itself?

**Comment by chrisvanpatten** [comment]
@boemedia Might you have the "Fix toolbar to Top" option active by mistake? Make sure it's not checked here:

![_untitled____add_new_post_ _chris_van_patten_ _wordpress_and_1__f___2___zsh__tmux_](https://user-images.githubusercontent.com/1231306/42724554-9bc754f8-8742-11e8-8de4-09f9964acb0a.png)

If that's checked, it would show the toolbar at the top of the screen, away from the block, as you described:

![_untitled____add_new_post_ _chris_van_patten_ _wordpress](https://user-images.githubusercontent.com/1231306/42724561-bc7fc4f0-8742-11e8-9954-60552d62d6ce.png)

If un-checked, the block controls should still be visible with the block itself:

![_untitled____add_new_post_ _chris_van_patten_ _wordpress](https://user-images.githubusercontent.com/1231306/42724570-d2a744e2-8742-11e8-805d-b3a12350105b.png)

**Comment by boemedia** [comment]
Thanks, @chrisvanpatten , I must be blondish... Totally forgot about that option being there. I must have been playing around with it with a previous update.

**Comment by pascalknecht** [comment]
@mtias I just checked the block variation implementation. I think it is great but I think it should not force the user to use a special kind of class system. Many people use BEM or other CSS styling methods to keep the CSS clean. The current implementation forces the user to use the class `.is-style-` which for me is too opiniated. Could we explore another way there?

**Comment by 0aveRyan** [comment]
To @pascalknecht's point, CSS classes aren't always used as style selectors. Common to target classes in JS for functionality -- `.is-style-init-vid-autoplay` or `.is-style-disable-vid-ads` would be confusing and verbose.

**Comment by galloppinggryphon** [comment]
Supporting what @pascalknecht said, I think the current system is too opinionated, and doesn't play nice with existing CSS frameworks, e.g. Bootstrap or UIKit. These already customized button styles and whatnot that you just can't use with the block editor as it is (please let me know if I'm wrong!) without, e.g. rolling your own buttons.

What about being able to directly register custom class names for each style? 
```
wp.blocks.registerBlockStyle( 'core/quote', {
    name: 'fancy-quote',
    label: 'Fancy Quote'
    classNames: 'pull-quote fancy-quote'
} );
```

**Comment by jorgefilipecosta** [comment]
Hi @mtias,
Regarding the feature "Custom User Styles", when the user (e.g: an admin) sets a new class as a style, should this block style be available to all users site-wide, or should this be an user-specific setting?

**Comment by chrisvanpatten** [comment]
> The user (e.g: an admin) sets a new class as a style, should this block style be available to all users site-wide, or should this be an user-specific setting?

I really like the idea of defaulting to globally-registered styles — it would allow a site creator to easily set up some "pre-sets" that their content creators can leverage later on.

**Comment by mtias** [comment]
@jorgefilipecosta site wide, depending on permissions.

**Comment by klihelp** [comment]
+1

**Comment by youknowriad** [comment]
A lot of improvements have been achieved for the block style variations. What's remaining here? Should we close this one?

**Comment by mtias** [comment]
I don't know if allowing a custom class name has been added, but it'd make sense to support it, even if discouraged, for multiple reasons.

**Comment by AL20** [comment]
Hi,
I don't if that's the right place to talk about it but I just updated the plugin to 7.9.0 and my buttons styles previews don't match my theme styles anymore :

![Uploading Capture d’écran 2020-04-16 à 17.42.11.png…]()

Every button preview in the sidebar have _.has-vivid-green-cyan-background-color_ class that override my theme custom style.

How can I fix that ?
Thanks,

**Comment by skorasaurus** [comment]
> I don't know if allowing a custom class name has been added, but it'd make sense to support it, even if discouraged, for multiple reasons.

It has not, there's an issue for it - https://github.com/WordPress/gutenberg/issues/11763

**Comment by coreyworrell** [comment]
Transferring #31360 to here as advised.

## What problem does this address?

Makes adding custom classes more user friendly for styling options that may not fit into the other options offered by Gutenberg already.

## What is your proposed solution?

Allow user to select from available classes that the theme offers along with adding custom. Similar to how [EditorsKit](https://jeffreycarandang.com/introducing-editorskit-block-styling-utility-classes/) plugin does it by using a token field. 

![](https://mk0jeffreycaranlitd1.kinstacdn.com/wp-content/uploads/2019/08/utility-classes.png)

Having the theme provide (via JSON and/or PHP) a list of classes (with labels) and possibly by group could make this very easy.

Utility classes for padding, margin, maximum widths, etc could take advantage of this. A label could be used with the underlying value being the class itself.

![Screenshot from 2021-04-29 13-19-23](https://user-images.githubusercontent.com/162595/116613531-9d9d9d80-a8ed-11eb-8d8b-28aee2d168f6.png)

**Comment by mikachan** [comment]
This is another issue that probably needs to be transferred to here: https://github.com/WordPress/gutenberg/issues/42297
